### ISSUE TITLE
Install Package Size Too Large

### ISSUE THREAD

**Issue by RaphaelS1** [alan-turing-institute/distr6#126]
Currently the installed package size is 5.2Mb, 0.2Mb over the limit. This is due to the R sub-directory (i.e. where the code is stored) being 4.4Mb. Unfortunately there is no easy way to work out which files are driving up the limit.

Using R's `object.size` gives the total package size as 1.9Mb.
Using `object_size` from `pryr` gives the total size as 117Mb, which is clearly wrong (I suspect that as this function takes into account environments, it accidentally registers R6 classes recursively within these).

To put this into perspective, running `object.size` on `mlr` gives a total size of 23Mb even though the true installed size is actually 5.1Mb.

75% of functions and classes are <= 5100 bytes (0.005Mb).

The bottom line is I have no idea what's going on but we can't submit until we have discovered the problem.

**Comment by fkiraly** [comment]
There's some bloat in the `references` part of the RDB, you can check by

```R
distr6RDS <- readRDS("whatever/distr6/R/distr6.rdx")
sum(sapply(distr6RDS$references,function(x){x[2]}))
```

According to the R internals manual these are serialized temp environments for adding named environments.
I'm not quite sure what causes this though.

**Comment by RaphaelS1** [comment]
That gives a total of 4.5Mb.... What exactly is it?

**Comment by RaphaelS1** [comment]
Or more importantly is that a problem with the code or with R's compression?

**Comment by fkiraly** [comment]
Don't know - while switching compression to bzip2 or xz might bring it under 5MB and make it CRAN-able, it probably wouldn't solve the root problem.

One would have to diagnose carefully what exactly sits in these bloated environments - I have no concrete suspicion.

**Comment by fkiraly** [comment]
(where the vague suspicion is that it's obviously R6 related, since R6 works primarily via environment encapsulation)

**Comment by RaphaelS1** [comment]
That's my suspicion too because I've looked at other packages which have physically more and larger functions and yet their install size is much less. I have a second suspicion that R62S3 is actually causing each reference to be doubled

**Comment by fkiraly** [comment]
Oh, are you running R62S3 *before* wrapping the package?

If so, that may be the issue. Could you try without?

More generally, I thought the design would require the user to run R62S3 themselves (as an option on load or manually) rather than it being run before packaging.

**Comment by fkiraly** [comment]
PS: I have a concrete suspicion now.

**Comment by RaphaelS1** [comment]
Can now confirm the problem is *not* R62S3. I have gone through and removed scripts one at a time until the installed size passed the test. Removing the entire 'SDistribution' script caused the package to run normally which tells us:

1. The problem isn't R62S3 as this doesn't run on SDistribution
2. As the package loads properly, nothing is being accidentally constructed in the function before being called

But! Every SDistribution child-class references the SDistribution documentation which in turn is a series of links to the S3 methods. So, I suspect the problem is that as there are 36 distributions and 11 kernels, each of which references the same parent class which in turn links to 40 different methods and we know that R is not clever enough to create pointer links but instead real links from each SDistribution to a method that means there are a grand total of 76 different help pages with 1,880 links between them. And as know it's the `references` part of the RDS, I suspect that is the problem.

**Comment by RaphaelS1** [comment]
> More generally, I thought the design would require the user to run R62S3 themselves (as an option on load or manually) rather than it being run before packaging.

Currently it is run automatically, on load. But this could be changed to run on attach (i.e. iff the user calls `library(distr6)`) or can be completely manual by the user. The only problem is that if the user calls it, all methods must be saved to the global environment, which can be cumbersome.

**Comment by fkiraly** [comment]
But since R62S3 is not the issue, that's actually fine.

**Comment by RaphaelS1** [comment]
So far I've tried:

1. Removing all sections that inherit from SDistribution - this didn't work
2. Removing all documentation from SDistribution - this didn't work
3. Not exporting SDistribution - this didn't work
4. Removing all public methods from SDistribution - didn't work (but did reduce total size to 5.0Mb)
5. Removing SDistribution class entirely - works

I should re-iterate the SDistribution class is a very small abstract class with three public methods, it has a small file size and doesn't inherit any documentation. 

I am lost

**Comment by RaphaelS1** [comment]
> But since R62S3 is not the issue, that's actually fine.

What's fine? If it loads automatically or if user loads it to GlobalEnv

**Comment by fkiraly** [comment]
Re. R62S3: fine to leave as-is, since you've isolated the problem to be somewhere else.

**Comment by fkiraly** [comment]
Where's `SDistribution` defined?

**Comment by fkiraly** [comment]
Found it

**Comment by fkiraly** [comment]
Question: in your fix 5, are you inheriting from `Distribution` instead of `SDistribution`?

**Comment by RaphaelS1** [comment]
No it's still set to inherit from `SDistribution`. R6 only looks for the inherited class when constructed, hence the code not breaking even though `SDistribution` is undefined when removed

**Comment by fkiraly** [comment]
Ah, so it's not actually a fix, since it breaks functionality?

**Comment by RaphaelS1** [comment]
Exactly it's just a way to try and figure out the problem

**Comment by RaphaelS1** [comment]
Update: Whilst deleting `SDistribution` somehow passes the package size, it doesn't actually decrease the saved RDB file at all. After systematically deleting files and checking the RDB size I've found that every implemented distribution comes out at 90KB after being processed and compressed by R (by comparison each individual distribution script is (4-10KB)). Therefore the large package size is down to each R6 class being massively inflated.

I am unsure how to solve this...

**Comment by RaphaelS1** [comment]
I have reduced the data-base size by 1.5Mb. I will mark this as closed once all checks passed.

Solution: Relates directly to issue #56. By having traits as variables they are defined in the public variables of a class, which is created in the package loading process. For every distribution these traits include 'type' which is an R6 class. By moving these to variables, the 'type' isn't constructed until the distribution is constructed.

**Comment by RaphaelS1** [comment]
Re-factored, down to 2.6Mb.
